### PR TITLE
For #831: Absolute volume control

### PR DESCRIPTION
**Pull request by Taknok**
## Implement an absolute volume control

An user can register 2 keys to control volume globally across all devices. A user muting a device with "normal" control cannot unmute with absolute control to prevent unwanted unmute.

The function `AbsoluteVolumeIncrement` and `AbsoluteVolumeDecrement` are in `App.xaml.xs` because I did not know if a new dedicated file is wanted for a such small addition or not. Tell me if you want to change the architecture to a cleaner structure.

The `absMuted` property is not set in the stream level because I did not see the value to set this low in the architecture. It is only needed by AudioSession. Again, I can change the location if wanted.

Compilation and run:
- [x] Debug
- [x] Release
- [x] VSDebug

Fixes: #831

### CODE REVIEW

**Comment by Taknok** [comment]
It seems that the build failed due to a dotnet SDK misconfiguration in the CI rather than the changes done by my code. Am I wrong ? :/

File: EarTrumpet/App.xaml.cs [hunk 1]
```diff
@@ -247,5 +249,44 @@ private SettingsCategoryViewModel CreateAddonSettingsPage(IEarTrumpetAddonSettin
...
+
+        private void AbsoluteVolumeDecrement()
+        {
+            foreach (var device in CollectionViewModel.AllDevices)
+            {
+                // if device is muted do noting
+                if (device.IsMuted)
+                {
+                    continue;
+                }
```
Comment (aloisdg) [inline EarTrumpet/App.xaml.cs:278]: If `AllDevices` is an Enumerable, we should be able to write:

           foreach (var device in CollectionViewModel.AllDevices.Where(d => !d.IsMuted))
           {
                // if device is not muted but will be muted by 
                bool wasMuted = device.IsMuted;
                // device.IncrementVolume(-2);
                device.Volume -= 2;
                // if device is muted by this absolute down
                // .IsMuted is not already updated
                if (!wasMuted == (device.Volume <= 0))
                {
                    device.absMuted = true;
                }
            }

File: EarTrumpet/App.xaml.cs [hunk 2]
```diff
@@ -247,5 +249,44 @@ private SettingsCategoryViewModel CreateAddonSettingsPage(IEarTrumpetAddonSettin
...
+                if (device.IsMuted && !device.absMuted)
+                {
+                    continue;
+                }
+
+                // in any case this device is not abs muted anymore
+                device.absMuted = false;
+                device.IncrementVolume(2);
+            }
+        }
```
Comment (aloisdg) [inline EarTrumpet/App.xaml.cs:268]: If `AllDevices` is an Enumerable, we should be able to write:

            foreach (var device in CollectionViewModel.AllDevices.Where(d => !d.IsMuted || d.absMuted))
            {
                // in any case this device is not abs muted anymore
                device.absMuted = false;
                device.IncrementVolume(2);
            }
            
Note that `!(a && !b)` is `!a || b`

**Comment by Taknok** [comment]
Thank you for the proposals !

**Comment by Taknok** [comment]
I will try to remove and reclone my project, because I have not this error on my computer.  This is strange, the resource "missing" is added in the commit https://github.com/File-New-Project/EarTrumpet/pull/1084/commits/0291aa0d544af4002c3c57367db7635cb9e9e718 ... along with the down version, which does not trigger any error. I compared both string and they are equals.

**Comment by riverar** [comment]
Merge from dev, will resolve the build failures since I made pipelines change there 😃

**Comment by Taknok** [comment]
No there is an issue... The branch is rebased in the latest dev. When I re-cloned the error code is returned at compilation, when re-saving the `Resources.resx` the compilation is working, but nothing is shown as modified in git status ... I tried CRLF/LF and nothing. I will investigate...

**Comment by riverar** [comment]
Oh I see, different error ```
D:\a\EarTrumpet\EarTrumpet\EarTrumpet\UI\Views\SettingsWindow.xaml(110,62): error MC3011: Cannot find the static member 'SettingsAbsoluteVolumeUpText' on the type 'Resources'
```

Will let you investigate :)

**Comment by Taknok** [comment]
By updating the `ResourceBuilder` from 16 to 17, the issue is gone in `Resources.Designer.cs` 47f65276126c2f0af8313fccc86510c5dc9471ee

**Comment by Taknok** [comment]
Any update on this ?

**Comment by riverar** [comment]
Tested this locally, works fine. Will go through PR now, thanks for your patience.

File: EarTrumpet/App.xaml.cs [hunk 3]
```diff
@@ -247,5 +249,33 @@ private SettingsCategoryViewModel CreateAddonSettingsPage(IEarTrumpetAddonSettin
         }
 
         private Window CreateMixerExperience() => new FullWindow { DataContext = new FullWindowViewModel(CollectionViewModel) };
+
+
+        private void AbsoluteVolumeIncrement()
+        {
+            foreach (var device in CollectionViewModel.AllDevices.Where(d => !d.IsMuted || d.absMuted))
```
Comment (riverar) [inline EarTrumpet/App.xaml.cs:256]: Nit: PascalCase public properties (and start with Is/Has)
```suggestion
            foreach (var device in CollectionViewModel.AllDevices.Where(d => !d.IsMuted || d.IsAbsMuted))
```

File: EarTrumpet/Properties/Resources.resx [hunk 4]
```diff
@@ -640,4 +638,10 @@ Open [https://eartrumpet.app/jmp/fixfonts] now?</value>
   <data name="ChooseAnAppHelpText" xml:space="preserve">
     <value>Apps shown above have played sound recently.</value>
   </data>
+  <data name="SettingsAbsoluteVolumeDownText" xml:space="preserve">
+    <value>Absolute Volume Down</value>
+  </data>
+  <data name="SettingsAbsoluteVolumeUpText" xml:space="preserve">
+    <value>Absolute Volume Up</value>
+  </data>
```
Comment (riverar) [inline EarTrumpet/Properties/Resources.resx:646]:
```suggestion
  <data name="SettingsAbsoluteVolumeDownText" xml:space="preserve">
    <value>Decrease volume for all devices</value>
  </data>
  <data name="SettingsAbsoluteVolumeUpText" xml:space="preserve">
    <value>Increase volume for all devices</value>
  </data>
```

File: EarTrumpet/UI/ViewModels/AudioSessionViewModel.cs [hunk 5]
```diff
@@ -8,12 +8,15 @@ namespace EarTrumpet.UI.ViewModels
     public class AudioSessionViewModel : BindableBase
     {
         private readonly IStreamWithVolumeControl _stream;
+        private bool _absMuted;
```
Comment (riverar) [inline EarTrumpet/UI/ViewModels/AudioSessionViewModel.cs:11]: Nit: 
```suggestion
        private bool _isAbsMuted;
```

**Review by riverar — CHANGES_REQUESTED** [review]

**Comment by riverar** [comment]
One thing you may want to look at (not a blocker for this PR): Holding the absolute shortcut down does not show changes in the UI and the app hangs while processing the events. This does not happen with our normal shortcut.

As soon as the minor PR nits are addressed I'll merge it in. Thanks again!

**Comment by Taknok** [comment]
On, I will try to fix this. I keep you updated.

**Comment by Taknok** [comment]
Converted to draft while adjusting shortcut processing

**Comment by riverar** [comment]
From build logs, just a friendly ping.
```
D:\a\EarTrumpet\EarTrumpet\EarTrumpet\App.xaml.cs(275,28): error CS1061: 'DeviceViewModel' does not contain a definition for 'absMuted' and no accessible extension method 'absMuted' accepting a first argument of type 'DeviceViewModel' could be found (are you missing a using directive or an assembly reference?) [D:\a\EarTrumpet\EarTrumpet\EarTrumpet\EarTrumpet_1lnm3op3_wpftmp.csproj
```

**Comment by Taknok** [comment]
Yes, I "drafted" the PR the time I get access to a computer to make a proper edit: replace all variables name, search the shortcut freeze and test a build.

**Comment by Taknok** [comment]
> Holding the absolute shortcut down does not show changes in the UI and the app hangs while processing the events.

What version of windows are you using ? Because I do not have this behavior :/

https://user-images.githubusercontent.com/12498020/168170545-72043281-7404-4193-aa57-9bef108cd0f0.mp4

**Review by riverar — APPROVED** [review]

**Comment by riverar** [comment]
Thanks again @Taknok! This is now in dev builds, will get it pushed out to the retail channel soon.

**Pull request merged**